staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,194 @@
+		}))
+		apfHandler.ServeHTTP(w, r)
+	})
+}
+
+const apfServerAddress = "127.0.0.1:6443"
+
+func newApfServer(apfHandler http.Handler) (*http.Server, error) {
+	listener, err := net.Listen("tcp", apfServerAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{"system:unauthenticated"},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	}), requestInfoFactory)
+
+	apfServerMux := http.NewServeMux()
+	apfServerMux.Handle("/api/v1/", handler)
+	apfServerMux.Handle("/metrics", legacyregistry.Handler())
+	apfServer := &http.Server{
+		Addr:    apfServerAddress,
+		Handler: apfServerMux,
+	}
+	go func() { apfServer.Serve(listener) }()
+	return apfServer, nil
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("http://%s/api/v1/namespaces?watch=true", apfServerAddress), http.StatusOK); err != nil {
+			// request should not be rejected
+			t.Error(err)
+		}
+	}()
+}
+
+func TestApfCanceledRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionCancel)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("http://%s/api/v1/namespaces/default", apfServerAddress), http.StatusTooManyRequests); err != nil {
+			t.Error(err)
+		}
+	}()
+
+	//wait metrics to be collected
+	time.Sleep(time.Second)
+
+	metrics, err := scrapeMetrics()
+	assert.NoError(t, err)
+	checkForExpectedMetrics(t, metrics, []string{
+		"apiserver_current_inqueue_requests",
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExecutingRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+	apfHandler := newApfDecisionHandler(decisionExecute)
+
+	server, err := newApfServer(apfHandler)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	go func() {
+		if err := expectHTTPGet(fmt.Sprintf("http://%s/api/v1/namespaces/default", apfServerAddress), http.StatusOK); err != nil {
+			t.Error(err)
+		}
+	}()
+
+	//wait metrics to be collected
+	time.Sleep(time.Second)
+
+	metrics, err := scrapeMetrics()
+	assert.NoError(t, err)
+
+	checkForExpectedMetrics(t, metrics, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_current_inqueue_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks_bucket",
+
+	})
+}
+
+func scrapeMetrics() (testutil.Metrics, error) {
```

Comment:
metrics gethering doesnt have to go thru the handler. in favor of `metrics.DefaultGatherer` to avoid unnecessary overhead..

Reply:
Done